### ISSUE TITLE
audio output compression

### ISSUE THREAD

**Issue by bmcfee** [librosa/librosa#70]
`librosa.output.write_wav` can dump wav output, but it would be nice to have a way to automatically run it through `ffmpeg` or `avconv` so we don't have intermediate wav files cluttering up the works.

**Comment by bmcfee** [comment]
I just took a look at PySoundFile 0.6, and it seems really nice.  It seems to do pretty much everything we want in terms of this issue, and issues #116 and #140. The one downside to switching from audioread to pysoundfile is that it's based on libsndfile, and thus does not support mp3.

I'd hate to lose mp3 support, but I also don't want to support multiple backends with differing functionaliy.  (That kinda defeats the purpose, right?)

Alternatively, the pysoundfile interface is almost plug-and-play with librosa as is, so maybe we can just let users mix the two on their own without any official integration?

What do people think?

**Comment by craffel** [comment]
> I'd hate to lose mp3 support, but I also don't want to support multiple backends with differing functionaliy. (That kinda defeats the purpose, right?)

Kind of, but this is basically what `audioread` does, so it's essentially what is currently being done although the onus is being placed on an external library.  I don't know of a way to avoid this currently (i.e. you either need to support multiple backends, or get everything except mp3 with libsndfile).

**Comment by bmcfee** [comment]
> Kind of, but this is basically what audioread does, so it's essentially what is currently being done although the onus is being placed on an external library

Yup.  I'm happy for audioread to do that for us, but it's also lacking a lot of functionality from pysndfile that would be nice to have.

OTOH, it probably makes sense to add pysndfile as an optional dependency anyway, and use it to replace our `output.write_wav` (falling back on scipy.io if you don't have psf).  If we're already going down that road, we may as well take advantage of the other functionality for input decoding.

> I don't know of a way to avoid this currently (i.e. you either need to support multiple backends, or get everything except mp3 with libsndfile).

Well, a moderate kludge solution would be to run everything through libsndfile, and fall back on a pipe to avconv whenever the input format isn't something sndfile-friendly.  (There might be a way to wrangle this out of audioread by way of psf and raw byte buffers, but that's probably slower than the avconv solution.)

**Comment by craffel** [comment]
> Well, a moderate kludge solution would be to run everything through libsndfile, and fall back on a pipe to avconv whenever the input format isn't something sndfile-friendly.

Yeah, I think this is the best solution I know of.

**Comment by bmcfee** [comment]
Having sat on this issue for years now, I think we should explicitly decide that audio coding is out of scope.

We can add a page to the documentation about compression via pysoundfile, but I don't think we should add any code.

Any objections?

**Comment by craffel** [comment]
I mean, it would be really useful, but if it's out of scope then it's out of scope.

**Comment by faroit** [comment]
for me audio coding would involve everything thats beyond wav/pcm, so using `ffmpeg` or `avconv` to write and read using lossy compression.
Under this definition, pysoundfile does not do compression but just read and write pcm wav files. The underlying `libsndfile` handles this way more flexible and robust than scipy.io.wavfile (especially with respect to bit depths), I would therefore propose to either replace the current output module with pysoundfile or remove the output completely and mention in the docs how to use pysoundfile or scipy.io.wavfile to write out wav files. Also it would then be even more consequent to add pysoundfile for reading audio or remove the audioread module as well, so librosa is would really be out of the audio codec business ;-)

**Comment by bmcfee** [comment]
> Under this definition, pysoundfile does not do compression but just read and write pcm wav files.

I don't understand this comment.  From the [pysoundfile docs](https://pysoundfile.readthedocs.io/en/0.8.1/#read-write-functions):

> Data can be written to the file using soundfile.write(), or read from the file using soundfile.read(). PySoundFile can open all file formats that libsndfile supports, for example WAV, FLAC, OGG and MAT files.

**Comment by faroit** [comment]
Ogg ist just a container not the actual codec. libsndfile does support [a bunch of codecs for ogg](https://github.com/erikd/libsndfile/blob/master/src/ogg.c) - mostly PCM based - but again this is probably not what users have in mind when they want to read or write compressed audio (like MP3, AAC, OPUS, Vorbis...).

**Comment by bmcfee** [comment]
Okay... but can't i output flac with pysoundfile as well?

**Comment by faroit** [comment]
So I was curious about this and double checked this issue. I revoke 1/2 of my statements above: 

* It looks like this depends on how `libsndfile` has been compiled and linked. A homebrew installed libsndfile on OSX does in fact allow you to use the VORBIS codec in OGG. Also it does allow you render flac. However this was different on other (windows machines). 

* One can query the available codecs from Pysoundfile by

```python
sf.available_subtypes('OGG')
{'VORBIS': u'Vorbis'}
```

```python
sf.available_subtypes('FLAC')
{'PCM_16': u'Signed 16 bit PCM',
'PCM_24': u'Signed 24 bit PCM',
'PCM_S8': u'Signed 8 bit PCM'}
```

* for VORBIS there seem to be no control about the codec's bit rate from `pysoundfile`. They are using the default, which is 128 kb/s (comparable to 128 kb/s MP3s)

* long story short: libsndfile _can_ handle compression for _some codecs_ and _some parameters_ if the libs are installed. It will not give you the common MP3, AAC and OPUS formats. But maybe FLAC is somethings that helps users to squeeze their audio pipelines a bit.

@bmcfee 
I would personally prefer to replace all wav reading and writing capabilities in librosa with pysoundfile. This would solve a lot of issues e.g. like handling 24bit wav correctly. Additionally this would give add light way to support compression like FLAC. 

> We can add a page to the documentation about compression via pysoundfile, but I don't think we should add any code.

so why not remove all read and write capabilities from librosa and add documentation (#420) how to use `pysoundfile`, `pyav` etc.?

**Comment by bmcfee** [comment]
> I would personally prefer to replace all wav reading and writing capabilities in librosa with pysoundfile. This would solve a lot of issues e.g. like handling 24bit wav correctly. Additionally this would add a light support for compression like FLAC.
> ...
> so why not remove all read and write capabilities from librosa and add documentation (#420) how to use pysoundfile, pyav etc.?

I'm fine with deprecating output (write) in favor of pysoundfile.

Replacing input is a trickier matter, for two reasons:

- The dependency chain needs to be maintained.  libsndfile is not installable via pip, and is not currently in conda.  We could add it to conda-forge, but that's a rather large set of responsibilities to take on.
- We'd lose mp3 support by relying solely on libsndfile.  I haven't conducted a proper user survey on the issue, but I expect this would be a significant problem for a lot of users.  I see no harm in adding documentation for how to use pysoundfile instead of `librosa.load`/audioread, but I don't think we should deprecate that functionality just yet.

**Comment by faroit** [comment]
> I'm fine with deprecating output (write) in favor of pysoundfile.

okay, I will add the examples next week, then.

> The dependency chain needs to be maintained. libsndfile is not installable via pip, and is not currently in conda. We could add it to conda-forge, but that's a rather large set of responsibilities to take on.

pysoundfile comes with dynamic libraries for [Windows and Mac](https://github.com/bastibe/libsndfile-binaries). I find this approach rather hacky, but it actually works quite nicely.

> We'd lose mp3 support by relying solely on libsndfile. I haven't conducted a proper user survey on the issue, but I expect this would be a significant problem for a lot of users. I see no harm in adding documentation for how to use pysoundfile instead of librosa.load/audioread, but I don't think we should deprecate that functionality just yet.

Ok, thats right... I should better endorse [MP3](http://www.fraunhofer.de) here ;-)
Also it's a pity that there is no audiowrite equivalent to audioread...

**Comment by bmcfee** [comment]
> pysoundfile comes with dynamic libraries for Windows and Mac. I find this approach rather hacky, but it actually works quite nicely.

I don't think we can get away with shipping those binaries as proper dependencies though.  Really, there ought to be conda-forge feeds for libsndfile and pysoundfile, so that we can reliably get the same behavior across platforms and have a properly integrated dependency chain.

This wouldn't necessarily be difficult to do, but it's more than I have time for myself right now or in the foreseeable future.  Maybe we can reach out to the pysdounfile folks and see if they'd be interested in taking it up?

**Comment by bmcfee** [comment]
Resolved by merging #480

**Comment by craffel** [comment]
Fraunhofer has released the mp3 patents:
https://www.iis.fraunhofer.de/en/ff/amm/prod/audiocodec/audiocodecs/mp3.html
Maybe if we are very lucky libsndfile will finally add mp3 support, then PySoundFile, then by extension librosa :)

**Comment by bmcfee** [comment]
Hot diggity. Is libsndfile on github so we can open an issue?

**Comment by craffel** [comment]
Indeed https://github.com/erikd/libsndfile
